web/packages/teleport/src/Console/DocumentSsh/DocumentSsh.tsx

```diff
@@ -77,29 +78,39 @@ export default function DocumentSsh({ doc, visible }: PropTypes) {
           afterClose={handleCloseFileTransfer}
           backgroundColor={colors.primary.light}
           transferHandlers={{
-            getDownloader: async (location, abortController) =>
-              getHttpFileTransferHandlers().download(
+            getDownloader: async (location, abortController) => {
+              const wanResponse = await auth.getAssertionResponseIfRequired({
```

Comment:
It might be too late to suggest this, but this is going to send an additional request every time, even if I never intend to use per-session MFA, right?

I wonder if it would be possible to get the information about MFA in some other way, perhaps as a return value to the function call which starts the SSH session and then pass it down to places which need to check whether MFA is required or not.

Reply:
there is another place where we use similar technique (for testing connections to a resource). i thought about this too and mentioned it but no one seemed to mind (no replies), so i figured maybe it was a pre-optimization question? 

that said, what do you think of this approach (it's similar to the way we handle it for web terminal):
 1) try to connect
 2) if an error returns, if the trace error is of type "access denied", check if MFA is required
      - if MFA is required, send back a 200 response with some flag set that says we need users mfa creds
      - if not, send back an error

Reply:
For this specific PR, I check if MFA is required and if they've sent an assertion response [here](https://github.com/gravitational/teleport/pull/22864/files#diff-ad4065e07d2573425dc5f7460e26dd324ffd893845ae1d3fe4b997113aba90cbR91). I could just send back an access denied + challenge to solve if thats the case. Not sure how it'd work with discovery

Reply:
I believe @rosstimothy did something similar for SSH - assume MFA is not required, if you get access denied, then check whether MFA is required and try again.

This way, most attempts don't incur any extra latency when MFA is not required.

Reply:
> I wonder if it would be possible to get the information about MFA in some other way, perhaps as a return value to the function call which starts the SSH session and then pass it down to places which need to check whether MFA is required or not.

After doing some tinkering, I think I prefer the original suggestion from @ravicious . I think the best place to make this happen is from the `useWebauthn` hook. There is a listener for `TermEvent.WEBAUTHN_CHALLENGE` which would happen on the SSH session init. We could set key in state like `mfaRequired` which would be set to true when we get that message and rely on that. The only issue I can see is if somehow a resource _doesnt_ require MFA anymore (I don't think this could happen) and we send a challenge assertion with a request that doesn't need it. It'll be skipped anyway so this doesn't prevent anything. Plus is a turbo edge case that I'm not even sure is an edge case.

If we already knew that mfa was required, then we just make the decision of which request to send client side only

No MFA required = file request
MFA required = challenge request -> file request


If we went the "check if access denied" path, it'd look more like
no MFA required = file request
MFA required = file request -> challenge request -> file request

My first thought was "just send a challenge in an access denied response body" but our api server doesn't give us access to the body. so I think the proposed solution is best.